lib/serial_shape.ml

```diff
@@ -0,0 +1,257 @@
+open Import
+
+module Conv = struct
+  type ('repr, 'true_type) t = {
+    from_repr : 'repr -> ('true_type, Rresult.R.msg) result;
+    to_repr : 'true_type -> 'repr;
+    equal : ('true_type -> 'true_type -> bool) option;
+    pp : 'true_type Fmt.t option;
+  }
+
+  let make ~from_repr ~to_repr ?equal ?pp () = { from_repr; to_repr; equal; pp }
+end
+
+type _ t =
+  | Sbool : bool t
+  | Sstring : string t
+  | Slist : 'a t -> 'a list t
+  | Spair : 'a t * 'b t -> ('a * 'b) t
+  | Choice2 : 'a t * 'b t -> [ `C1 of 'a | `C2 of 'b ] t
+  | Choice3 : 'a t * 'b t * 'c t -> [ `C1 of 'a | `C2 of 'b | `C3 of 'c ] t
+  | Conv : ('repr, 'true_type) Conv.t * 'repr t -> 'true_type t
+
+let bool = Sbool
+let string = Sstring
+let list s = Slist s
+let pair s s' = Spair (s, s')
+let choice2 s s' = Choice2 (s, s')
+let choice3 s s' s'' = Choice3 (s, s', s'')
+let conv conv s = Conv (conv, s)
+
+let rec shallow_description : type a. a t -> string = function
+  | Slist _ -> "a list"
+  | Spair _ -> "a pair"
+  | Sstring -> "a string"
+  | Sbool -> "a boolean"
+  | Choice2 (s, s') ->
+      let d = shallow_description s in
+      let d' = shallow_description s' in
+      Printf.sprintf "%s or %s" d d'
+  | Choice3 (s, s', s'') ->
+      let d = shallow_description s in
+      let d' = shallow_description s' in
+      let d'' = shallow_description s'' in
+      Printf.sprintf "%s, %s or %s" d d' d''
+  | Conv (_, s) -> shallow_description s
+
+let rec from_opam_val :
+    type a. a t -> OpamParserTypes.FullPos.value -> (a, Rresult.R.msg) result =
+ fun shape value ->
+  let open Result.O in
+  let parse_error () =
+    let expected = shallow_description shape in
+    Opam.Pos.unexpected_value_error ~expected value
+  in
+  match (shape, value) with
+  | Sbool, { pelem = Bool b; _ } -> Ok b
+  | Sstring, { pelem = String s; _ } -> Ok s
+  | Spair (s, s'), { pelem = List { pelem = [ v; v' ]; _ }; _ } ->
+      let* fst = from_opam_val s v in
+      let* snd = from_opam_val s' v' in
+      Ok (fst, snd)
+  | Slist s, { pelem = List _; _ } ->
+      let* l = Opam.Value.List.from_value (from_opam_val s) value in
+      Ok l
+  | Conv (conv, s), value -> (
+      let* repr = from_opam_val s value in
+      let res = conv.from_repr repr in
+      match res with
+      | Ok _ as res -> res
+      | Error (`Msg msg) -> Opam.Pos.value_errorf ~value "%s" msg)
+  | Choice2 (s, s'), value -> (
+      match from_opam_val s value with
```

Comment:
Using lists here is difficult because the return type differs for each case. There are mechanism to deal with GADTs in such cases but it's pretty complex and can get rather messy just to avoid a 3 branch pattern matching which is quite straightforward.
if we have to handle more of those in the future we can consider refactoring it but unless you are strongly opposed to it, I'd suggest we stick with this for now as it's probably the simpler solution.